File: agent/sdk/typescript/examples/budget-constrained-spending.ts
import { Client, type IdempotentMutation } from "../src/index.js";

export interface BudgetRequest { readonly limit: bigint; readonly enforcement: "ProtocolBudget" | "DaemonLimit" }

export async function createBudget(
  client: Client,
  request: IdempotentMutation<BudgetRequest>,
): Promise<unknown> {
  if (request.operation.limit <= 0n) throw new RangeError("budget_limit");
  return client.call("budget.create", request);
}

File: agent/sdk/python/examples/budget_constrained_spending.py
from dataclasses import dataclass
from typing import Literal

from layerx_sdk import Client, IdempotentMutation


@dataclass(frozen=True)
class BudgetRequest:
    limit: int
    enforcement: Literal["ProtocolBudget", "DaemonLimit"]


def create_budget(client: Client, request: IdempotentMutation[BudgetRequest]) -> object:
    if request.operation.limit <= 0:
        raise ValueError("budget_limit")
    return client.call("budget.create", request)
